### PR TITLE
NGMesh and ArchVis Plugins

### PR DESCRIPTION
**Pull request by invor**
NGMesh Plugin: A modern renderer for arbitrary meshes. All data (shader, mesh, draw commands, shader parameters) is prepared and supplied by the data source. An example/template for such a datasource will be included.

ArchVis Plugin: Plugin for scientific architectural visualizations in the context of SFB1244.

### CODE REVIEW

File: plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+	class NG_MESH_API CallNGMeshRenderBatches : public megamol::core::AbstractGetData3DCall
+	{
+		public:
+
+			class NG_MESH_API RenderBatchesData
+			{
+			public:
+
+				static uint32_t shaderUpdateBit()			{ return 0x1; }
```
Comment (reinago) [inline plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h]: why? plus it's not even inline. enums work okay, don't they?

File: plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+					delete raw_buffer;
+					raw_buffer = new uint8_t[allocated_byte_size];
+					std::memcpy(raw_buffer, rhs.raw_buffer, used_byte_size);
+
+					// set individual pointers to new values
+					size_t base_offset = 0;
+					size_t offset = 0;
+					for (int i = 0; i < batch_cnt; ++i)
+					{
+						shader_prgms = reinterpret_cast<ShaderPrgmData*>(raw_buffer);
```
Comment (reinago) [inline plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h]: no offsets?

File: plugins/ng_mesh/src/NGMeshRenderer.cpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+	for (auto& render_batch : m_render_batches)
+	{
+		render_batch.shader_prgm->Enable();
+
+		render_batch.mesh_shader_params->bind(0);
+		render_batch.mtl_shader_params->bind(1);
+
+		render_batch.draw_commands->bind();
+		render_batch.mesh->bindVertexArray();
+		/*
```
Comment (reinago) [inline plugins/ng_mesh/src/NGMeshRenderer.cpp]: ?

**Review by reinago — CHANGES_REQUESTED** [review]
- it does not work yet, right? why merge now?
- Readme.md please if you merge
- I don't really like Pascal-style brackets ;)

File: plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h (hunk 2 above)
Comment (invor) [reply plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h]: Fixed in f8bfeb7

File: plugins/ng_mesh/src/NGMeshRenderer.cpp (hunk 3 above)
Comment (invor) [reply plugins/ng_mesh/src/NGMeshRenderer.cpp]: What seems to be the problem? glMultiDrawElementsIndirect requires draw commands buffer (SSBO in this case) + geometry to work with given by a vertex array object.

File: plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h (hunk 1 above)
Comment (invor) [reply plugins/ng_mesh/include/ng_mesh/CallNGMeshRenderBatches.h]: Edit: Fixed in 9863f3b

**Review by tobiasrau — CHANGES_REQUESTED** [review]
- [x] Should be merged with the actual master.
 - [x] Build checks missing
 - [x] glload is deprecated
 - [x] externals should be stored in a external subdirectory in the plugin, not in the plugins `src` directory
 - [ ] `Readme.md` is empty

**Comment by tobiasrau** [comment]
Cleaned and splitted into #314 and #315.

**Pull request closed without merging**